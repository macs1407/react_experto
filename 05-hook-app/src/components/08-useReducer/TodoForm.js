import React from 'react';
import { useForm } from '../../hooks/useForm';

export const TodoForm = ({handleAddTodo}) => {

    const [{description}, handleInputChange, reset] = useForm({
        description: ''
    });

    // Agregar TUDU
    const handleSubmit = (e)=>{
        e.preventDefault();

        if(description.trim().length <= 1) return;
        
        const nuevoTudu = {
            id: new Date().getTime(),
            desc:description,
            done:false
        };

        // Enviar la funcion que cambia el useReducer
        handleAddTodo(nuevoTudu);
        // Resetar el formuario
        reset();
    }

    return (
        <>
            <h4>Agregar tarea</h4>
            <hr />
            <form onSubmit={handleSubmit}>
                <input 
                    type="text"
                    name="description"
                    placeholder="tarea"
                    autoComplete="off"
                    className="form-control"
                    onChange={handleInputChange}
                    value={description}
                />
                <button className="btn btn-outline-primary mt-1 btn-block" type="submit">Agregar</button>
            </form>  
        </>
    )
}
