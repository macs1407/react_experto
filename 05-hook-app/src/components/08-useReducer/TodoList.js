import React from 'react'
import { TodoListItem } from './TodoListItem'

export const TodoList = ({tudus, handleDelete, handleToogle}) => {

    return (
        <ul className="list-group list-group-flush">
        {
            tudus.map((tarea, indice)=>(
                <TodoListItem 
                    key={tarea.id}
                    todo={tarea} 
                    index={indice} 
                    handleDelete={handleDelete} 
                    handleToogle={handleToogle}/>
            ))
        }                
        </ul>
    )
}
