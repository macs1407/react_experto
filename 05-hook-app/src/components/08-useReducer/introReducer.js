// Crear state inicial
const initialState = [{
    id:1,
    todo:'Hacer al aplicacion de shardMusc',
    donde:false,
}];
// Reducer basico
// el reducer recibe dos argumentos, el state y la accion que va a cambiar el state
const todoReducer = (state = initialState, action)=>{
    // Acciones a realizar sobre el state
    if(action?.type === 'agregar'){
        return [...state, action.payload];
    }
    // Siempre se tiene que retornar un state
    return state;
}

// Inicializar
let todos = todoReducer();

// Crear un nuevo objeto
const nuevoTodo = {
    id:2,
    todo:'Seguir estudiando',
    donde:false,
};

// Para enviar un nuevo todo, se crea un objeto con 
// Type: accion a ejecutar
// payload: objeto que se envia
const agregarAccion = {
    type:'agregar',
    payload: nuevoTodo
}

// Enviar un nuevo objeto
todos = todoReducer(todos, agregarAccion);

// Imprimir la lista de tareas
console.log(todos);