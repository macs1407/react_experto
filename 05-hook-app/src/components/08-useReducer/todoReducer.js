export const todoReducer = (state = [], action)=>{
    switch(action.type){
        case 'add':
            return [...state, action.payload];
        // Regresa un nuevo arreglo menos el id que se envia
        case 'delete':
            return state.filter(tudu=> tudu.id !== action.payload);
        // Forma larga
        case 'toogle-old':
            return state.map(todo=>{
                // Es tudu que toca cambiar
                if(todo.id === action.payload){
                    return {
                        ...todo,
                        done: !todo.done
                    }
                } else {
                    return todo
                }
            });
        // Forma corta
        case 'toogle':
                return state.map(todo=>
                    (todo.id === action.payload) ? {...todo, done: !todo.done} : todo
                );
        default:
            return state;
    }
}