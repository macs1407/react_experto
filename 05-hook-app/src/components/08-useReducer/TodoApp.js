import React, { useReducer, useEffect } from 'react';
import './estilos.css';
import { todoReducer } from './todoReducer';
import { TodoList } from './TodoList';
import { TodoForm } from './TodoForm';

const init = ()=>{
    // Obtener los tudus del localStorage
    return JSON.parse(localStorage.getItem('tudus')) || [];
}

export const TodoApp = () => {
    // El dispatch es un funcion que se le manda una accion y se encarga de modificar el state
    // El init es una funcion que se ejecuta para establecer el estado inicial del reducer
    const [tudus, dispatch] = useReducer(todoReducer, [], init);


    // Eliminar un tudu
    const handleDelete = (tuduId)=>{
        const eliminarTudu = {
            type: 'delete',
            payload: tuduId
        }
        dispatch(eliminarTudu);
    }

    const handleToogle = (tuduId)=>{
        dispatch({
            type:'toogle',
            payload:tuduId
        })
    }

    const handleAddTodo = (newTodo)=>{
        dispatch({
            type:'add',
            payload:newTodo
        })
    }

    // Se ejecuta por lo menos una vez
    // Si los tudus cambian se va avolver a grebar en el localStorage
    useEffect(()=>{
        localStorage.setItem("tudus", JSON.stringify(tudus));
    },[tudus]);

    return (
        <div>
            <h1>Tareas App / cantidad {tudus.length}</h1>
            <hr />
            <div className="row">
                <div className="col-7">
                    <TodoList tudus={tudus} handleDelete={handleDelete} handleToogle={handleToogle}/>
                </div>
                <div className="col-5">
                   <TodoForm handleAddTodo={handleAddTodo}/>
                </div>
            </div>
           
        </div>
    )
}
