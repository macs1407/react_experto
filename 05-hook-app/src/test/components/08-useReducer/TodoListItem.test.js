import React from 'react';
import '@testing-library/jest-dom';
import {shallow} from 'enzyme';
import { TodoListItem } from '../../../components/08-useReducer/TodoListItem';
import { demoTodos } from '../../fixtures/demoTodos';

describe('Probando TodoListItem', () => {

    const handleDelete = jest.fn();
    const handleToogle = jest.fn();

    const wraper = shallow(<TodoListItem 
        todo={demoTodos[0]} 
        index={0} 
        handleDelete={handleDelete}
        handleToogle={handleToogle}/>);
    
    test('Debe de mostrarse correctamente ', () => {
        
        expect(wraper).toMatchSnapshot(); 
    }); 

    test('Debe de llamar al handleDelete ', () => {
        wraper.find('button').simulate('click')
        expect(handleDelete).toHaveBeenCalledWith(demoTodos[0].id) 
    });

    test('Debe de llamar al handleToogle ', () => {
        wraper.find('p').simulate('click')
        expect(handleToogle).toHaveBeenCalledWith(demoTodos[0].id) 
    });
     
    test('Debe de mostrar el texto correctamente ', () => {
        const parrafo = wraper.find('p').text();
        console.log(parrafo)
        expect(parrafo.trim()).toBe('1. '+demoTodos[0].desc); 
    });
    
})

