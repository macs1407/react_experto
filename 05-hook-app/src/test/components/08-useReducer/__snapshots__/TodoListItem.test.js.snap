// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Probando TodoListItem Debe de mostrarse correctamente  1`] = `
<li
  className="list-group-item"
  key="1"
>
  <p
    className="false"
    onClick={[Function]}
  >
    1
    . 
    aprender react
  </p>
  <button
    className="btn btn-danger"
    onClick={[Function]}
  >
    Borrar
  </button>
</li>
`;
